app/models/host/discovered.rb

```diff
@@ -137,33 +137,49 @@ def proxied?
     subnet.present? && subnet.discovery.present?
   end
 
-  def proxy_url
-    proxied? ? subnet.discovery.url + "/discovery/#{self.ip}" : "https://#{self.ip}:8443"
+  def proxy_url(ip)
+    proxied? ? subnet.discovery.url + "/discovery/#{ip}" : "https://#{ip}:8443"
   end
 
   def refresh_facts
-    facts = ::ForemanDiscovery::NodeAPI::Inventory.new(:url => proxy_url).facter
+    facts = ::ForemanDiscovery::NodeAPI::Inventory.new(:url => proxy_url(self.ip)).facter
     self.class.import_host facts
     import_facts facts
   rescue => e
     ::Foreman::Logging.exception("Unable to get facts from proxy", e)
     raise ::Foreman::WrappedException.new(e, N_("Could not get facts from proxy %{url}: %{error}"), :url => proxy_url, :error => e)
   end
 
-  def reboot
-    resource = ::ForemanDiscovery::NodeAPI::Power.service(:url => proxy_url)
+  def reboot(ip = facts["discovery_bootip"] || facts["ipaddress"] || self.ip)
```

Comment:
I think this would be more readable as a kwarg, something like this:
```suggestion
  def reboot(proxy_ip: facts["discovery_bootip"] || facts["ipaddress"] || self.ip)
```

Reply:
Well not so much for the `kexec` method down below. I slightly lean towards keeping regular argument.